src/NLSModels.jl

```diff
@@ -93,8 +93,8 @@ include("feasibility_residual.jl")
 
 Computes F(x), the residual at x.
 """
-function residual(nls :: AbstractNLSModel, x :: AbstractVector)
-  Fx = zeros(nls_meta(nls).nequ)
+function residual(nls :: AbstractNLSModel, x :: AbstractVector{T}) where T <: Real
+  Fx = zeros(T, nls_meta(nls).nequ)
```

Comment:
Is it necessary to parametrize? Couldn't we write `Fx = zeros(eltype(x), ...)`?

Reply:
It works, but I actually don't know which is better.

Reply:
I've made a new commit which uses only `eltype(x)`.
From what I tested:
- Using `x :: AbstractVector{T}` in some functions creates a cascade of using of `T` in other functions, like `SlackModel`, `Feasibility` stuff and `QNModels`.
- Using `T <: Real` also reduces the possibilities of what can be passed to function. We may want that, but it so then we would have to define it for every other function for consistency.